edge/pkg/metamanager/metamanager.go

```diff
@@ -67,6 +68,10 @@ func (m *metaManager) Start() {
 	if metaserverconfig.Config.Enable {
 		imitator.StorageInit()
 		go metaserver.NewMetaServer().Start(beehiveContext.Done())
+
+		if metaserverconfig.Config.Debug {
+			go helper.AddCRD()
```

Comment:
I'm not sure why it should be executed when `debug` is turned on.
This method adds some istio related CRD resources. What scenarios are they mainly used for?

Reply:
This pr is mainly to do resource-kind conversion for some irregular resources. This function uses Gateway as an example to test this pr. I need to put this resource in sqlite in advance, so I only need to open it in the `debug` mode.

Reply:
Thank you for your explanation. Can I think this is a test for resource conversion?
So, I think it should not be put in the actual service, it should be more reasonable to test in the test framework, what do you think?

Reply:
yes, it is a test for resource conversion.
Now I do test it in `edge/test/integration/metaserver/metaserver_test.go` (can this be understood as a test framework?) 
Do you mean to place these resources in metaserver_test.go？

Reply:
Yes, I think it should be.

Reply:
I tried this way, but I found some problems.
When the resource is inserted through the function `imitator.DefaultV2Client.InsertOrUpdateObj()`, the corresponding key will be generated. https://github.com/kubeedge/kubeedge/blob/5165135de8c2aab5a23d04a1e25c4e8f904a58c1/pkg/metaserver/key.go#L27-L54
During this process, the `UnsafeKindToResource()`  will be used. However, the map `CRDKindToResource[][]` of the `edgecore` process cannot be accessed by the `integrationtest` process. 
That is to say, if I choose to call `imitator.DefaultV2Client.InsertOrUpdateObj()` to insert data in the `integrationtest` process, the correct key will not be generated. 
And it is meaningless to write the correspondence of kind and resource in my `integrationtest` process directly. I want to verify whether the `UnsafeKindToResource()` in the `edgecore` process works, so I must insert data in the `edgecore` process.

Reply:
Let me try to understand that you want to test getting some CRD resources in the running edgecore?

Reply:
yes, UpdateCrdMap() will get crd regularly to update the CRDKindToResource[][] and CRDResourceToKind[][] map, and  when we call `imitator.DefaultV2Client.InsertOrUpdateObj()`, it will use the info in CRDKindToResource[][] to generate the key of the corresponding resource to store in sqlite.

Reply:
Thanks for your explanation.
I think this whole test action should be completed in the `ginkgo` framework. When `cloudcore` and `edgecore` are running, try to create crd resources in the `kind`, and then obtain them through the `metaserver` of `edgecore`.

Can this test the results you want?

Reply:
Can I understand it as a system test？ I can deploy cloudcore and edgecore, and then `kubectl apply -f` a crd to test it, but I don’t know how to implement such a process through some `_test.go` file. Can you explain? or show me some examples, thx:)

Reply:
You can find the documentation for ginkgo here [https://github.com/onsi/ginkgo](https://github.com/onsi/ginkgo).
And you can write the tests in `tests/e2e`.
In this way, when the `e2e_test` step is executed by `github action`, the test case you wrote will be run.

Reply:
well, I am trying to solve it, but I have encountered some problems. 
When `make e2e`, I found that the program was looping here all the time.
https://github.com/kubeedge/kubeedge/blob/e0100a0c280002ef6b9b4f456a5de1c2a2026236/hack/local-up-kubeedge.sh#L243-L251
![image](https://user-images.githubusercontent.com/34911960/134460426-1fd6aa81-4527-49fe-ba99-b95a6048c03b.png)
I tried it on different virtual machines and they got stuck here. What is the reason? It doesn't seem to be the case when testing in github? o(╥﹏╥)o

Reply:
+1, we should not add the test code here.